comet/broker.py

```diff
@@ -513,33 +513,31 @@ async def gather_update(ts, roots):
     given roots that were registered after the given timestamp.
     """
     update = dict()
-    async with lock_datasets:
-        for root in roots:
+    for root in roots:
+        async with lock_datasets:
             # Get both dicts from redis concurrently:
             keys, tree = await asyncio.gather(
                 redis.execute("hget", "datasets_of_root_keys", root),
                 redis.execute("hget", "datasets_of_root", root),
             )
 
-            keys = reversed(json.loads(keys))
-            tree = list(reversed(json.loads(tree)))
-
-            # The nodes in tree are ordered by their timestamp from new to
-            # old, so we are done as soon as we find an older timestamp than
-            # the given one.
-            tasks = []
-            for n, k in zip(tree, keys):
-                if k < ts:
-                    break
-                tasks.append(
-                    asyncio.ensure_future(redis.execute("hget", "datasets", n))
-                )
-
-            if tasks:
-                # Wait for all concurrent tasks (gather keeps their order)
-                tasks = await asyncio.gather(*tasks)
-                # put back together the root ds IDs and the datasets
-                update.update(dict(zip(tree, [json.loads(task) for task in tasks])))
+        keys = reversed(json.loads(keys))
+        tree = list(reversed(json.loads(tree)))
+
+        # The nodes in tree are ordered by their timestamp from new to
+        # old, so we are done as soon as we find an older timestamp than
+        # the given one.
+        tasks = []
+        for n, k in zip(tree, keys):
+            if k < ts:
+                break
+            tasks.append(asyncio.ensure_future(redis.execute("hget", "datasets", n)))
+
+        if tasks:
+            # Wait for all concurrent tasks (gather keeps their order)
+            tasks = await asyncio.gather(*tasks)
+            # put back together the root ds IDs and the datasets
+            update.update(dict(zip(tree, [json.loads(task) for task in tasks])))
```

Comment:
Presumably you could split this up into a loop which produces the list of datasets with matching timestamps, and then using a single `hmget` call to get all of them at once.

Reply:
I can make this change, but we would have to repeat the live test from this morning.